File: 3453-separate-squares-i/3453-separate-squares-i.py
def calcArea(squares: List[List[int]], y: float) -> float:
    area = 0
    for sq in squares:
        area += sq[2] * max(0, min(sq[2], y - sq[1]))
    return area

class Solution:
    def separateSquares(self, squares: List[List[int]]) -> float:
        totArea = 0
        minY = float('inf')
        maxY = 0
        for sq in squares:
            totArea += sq[2] * sq[2]
            minY = min(minY, sq[1])
            maxY = max(maxY, sq[1] + sq[2])
        eps = 10**-5
        while abs(maxY - minY) > eps:
            mid = (maxY + minY) / 2
            area = calcArea(squares, mid)
            if area * 2 >= totArea:
                maxY = mid
            else:
                minY = mid
        return maxY

        

File: 3453-separate-squares-i/3453-separate-squares-i.cpp
class Solution {
public:
    double separateSquares(vector<vector<int>>& squares) {
        double totArea = 0;
        double minY = INT_MAX;
        double maxY = 0;
        double eps = pow(10, -5);
        for(int i = 0; i < squares.size(); i++)
        {
            totArea += (double)squares[i][2] * squares[i][2];
            minY = min(minY, (double)squares[i][1]);
            maxY = max(maxY, (double)squares[i][1] + squares[i][2]);
        }
        while(abs(maxY - minY) > eps)
        {
            double mid = (minY + maxY) / 2;
            double area = calcArea(squares, mid);
            if(area * 2 >= totArea)
            {
                maxY = mid;
            }
            else
            {
                minY = mid;
            }
        }
        return maxY;
    }
private:
    double calcArea(vector<vector<int>>& squares, double y)
    {
        double area = 0;
        for(int i = 0; i < squares.size(); i++)
        {
            area += (double)squares[i][2] * max(0.0, min((double)squares[i][2], y - squares[i][1]));
        }
        return area;
    }
};